homeassistant/components/wled/config_flow.py

```diff
@@ -0,0 +1,134 @@
+        name = self.context.get(CONF_NAME)
+        return self.async_show_form(
+            step_id="zeroconf_confirm",
+            description_placeholders={"name": name},
+            errors=errors or {},
+        )
+
+    async def _handle_config_flow(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        source = self.context.get("source")
+
+        if user_input is None:
+            if source == SOURCE_ZEROCONF:
+                return await self._show_confirm_dialog()
+            return await self._show_setup_form()
+
+        if source == SOURCE_ZEROCONF:
+            # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+            user_input[CONF_HOST] = self.context.get(CONF_HOST)
+
+        errors = {}
+        session = async_get_clientsession(self.hass)
+        wled = WLED(user_input[CONF_HOST], loop=self.hass.loop, session=session)
+
+        try:
+            device = await wled.update()
+            mac_address = device.info.mac_address
+        except WLEDConnectionError:
+            if source == SOURCE_ZEROCONF:
+                return self.async_abort(reason="connection_error")
+            errors["base"] = "connection_error"
+            return await self._show_setup_form(errors)
+
+        # Check if already configured
+        for entry in self._async_current_entries():
+            if entry.data[CONF_MAC] == mac_address:
+                # This mac address is already configured
+                return self.async_abort(reason="already_configured")
+
+        title = user_input[CONF_HOST]
+        if source == SOURCE_ZEROCONF:
+            # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+            title = self.context.get(CONF_NAME)
+
+        return self.async_create_entry(
+            title=title, data={CONF_HOST: user_input[CONF_HOST], CONF_MAC: mac_address}
+        )
+
+    async def async_step_user(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by the user."""
+        return await self._handle_config_flow(user_input)
+
+    async def async_step_zeroconf_confirm(
+        self, user_input: ConfigType = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by zeroconf."""
+        return await self._handle_config_flow(user_input)
+
+    async def async_step_zeroconf(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle zeroconf discovery."""
+        if user_input is None:
+            return self.async_abort(reason="connection_error")
+
+        # Hostname is format: wled-livingroom.local.
+        host = user_input["hostname"].rstrip(".")
+        name, _ = host.rsplit(".")
+
+        # Set up connection to WLED to find its unique identifier
+        session = async_get_clientsession(self.hass)
```

Comment:
This part looks very similar to the code in the `_handle_config_flow`. Can we consolidate eg by having another helper method that accepts host as parameter?

I also don't understand why we need to set up the connection first here and then again in `_handle_config_flow` if we're setting up via zeroconf.

Reply:
Because zeroconf has to figure out if the found device is unique. This can only be checked by contacting the API at this point.
WLED has a pending change for this, however, that is not available right now.

Reply:
I think the flow instance will be garbage collected eventually. We only store it during the flow progress and pop it when creating config entry or aborting the flow.

Reply:
I've started doing it, however, I aborted that process.
The main reasoning, is that with the next WLED update, the unique identifier will be part of the zeroconf data, which makes this connection not needed at all.

I've centralized it the connection nevertheless.